dataflowAPI/rose/semantics/Registers.C

```diff
@@ -643,6 +644,31 @@ RegisterDictionary::print(std::ostream &o) const {
 //    }
 //    return regs;
 //}
+//
+/** AMDGPU registers.
+ * There are a total of 104 scalar general purpose registers each 32 bits wide. Each of these registers can be addressed as its 32-bit or 64-bit form */
+const RegisterDictionary *
+RegisterDictionary::dictionary_amdgpu() {
```

Comment:
Leaky singletons are my least favorite singleton. Could we change this to

    RegisterDictionary const& RegisterDictionary::dictionary_amdgpu() {
        static std::once_flag initialized;
        static RegisterDictionary regs;

        std::call_once(...)

        return regs;
    }

or

    RegisterDictionary const* RegisterDictionary::dictionary_amdgpu() {
        static std::once_flag initialized;
        static RegisterDictionary regs;

        std::call_once(...)

        return &regs;
    }

if you need to preserve the pointer semantics of the return value? Both versions are safe in terms of lifetimes and threading.

Reply:
This part I'm just following what was done with armv8.
I can probably fix the code to use the second pattern.
Can you explain to me what is problem with this code ? 
Like I'm not that familiar with C++, so  what would be the difference between the original version and the version you've suggested.

Reply:
I'm definitely not faulting you for using the existing code as a template here, even though the method is sub-optimal. I just want to make sure we are improving the code as we update it. The leaky single pattern (i.e., never freeing the instance) is only useful when there is a problem in ordering of destruction of static objects. That's not an issue here, so we can use better patterns that don't cause fault positives when run under a memory analysis tool (e.g., valgrind).

From the C++ side of things, C++11 added thread safety to function-local static variables. This means that `regs` is only constructed once (when the function is first called) and that initialization is thread-safe. This wasn't the case when the previous code was written (i.e., C++98). This new guarantee allows us to use a local variable with static storage duration instead of storing it on the heap. The call to `call_once` is still necessary because we are doing some non-trivial work that should only be done once. Does that help at all? It's one of the more technical changes made to the language, so I'm not surprised you are unfamiliar with it.

Reply:
So this part of the code is removed as an effort to separate out the work on dataflowAPI. I'll mark it as resolved here.